Teach compiler about partitioned bindings

Keno opened the pull request:
This commit teaches to compiler to update its world bounds whenever it looks at a binding partition, making the compiler sound in the presence of a partitioned binding. The key adjustment is that the compiler is no longer allowed to directly query the binding table without recording the world bounds, so all the various abstract evaluations that look at bindings need to be adjusted and are no longer pure tfuncs. We used to look at bindings a lot more, but thanks to earlier prep work to remove unnecessary binding-dependent code (#55288, #55289 and #55271), these changes become relatively straightforward.

Note that as before, we do not create any binding partitions by default, so this commit is mostly preperatory.

Keno committed:
Teach compiler about partitioned bindings

This commit teaches to compiler to update its world bounds whenever
it looks at a binding partition, making the compiler sound in the
presence of a partitioned binding. The key adjustment is that the
compiler is no longer allowed to directly query the binding table
without recording the world bounds, so all the various abstract
evaluations that look at bindings need to be adjusted and are no
longer pure tfuncs. We used to look at bindings a lot more, but
thanks to earlier prep work to remove unnecessary binding-dependent
code (#55288, #55289 and #55271), these changes become relatively
straightforward.

Note that as before, we do not create any binding partitions by
default, so this commit is mostly preperatory.

Keno commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -2316,6 +2316,173 @@ function abstract_throw_methoderror(interp::AbstractInterpreter, argtypes::Vecto
+function abstract_eval_get_binding_type(interp::AbstractInterpreter, sv::AbsIntState, argtypes::Vector{Any})
+    if length(argtypes) == 3
+        return abstract_eval_get_binding_type(interp, sv, argtypes[2], argtypes[3])
+    elseif !isvarargtype(argtypes[end]) && length(argtypes) > 4
```
Good catch, thanks.

Keno committed:
Update base/compiler/abstractinterpretation.jl

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>

Keno committed:
Performance hack abstract_eval_globalref_type

Keno committed:
Fix some atomics effect models

Keno commented:
@nanosoldier `runbenchmarks()`

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56299#issuecomment-2449071080) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/cc771a4_vs_2a24b8f/report.md).

Keno commented:
Master was the wrong comparison, since this PR was outdated. Results looked fine, except for the memory regressions in the inference tests. Let's re-run those to make sure that's actually something that regressed here rather than something that got fixed on master.


@nanosoldier runbenchmarks("inference", vs=":master")

Keno commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

Keno committed:
Add a missing GC root in constant declaration

As pointed out in https://github.com/JuliaLang/julia/pull/56224#discussion_r1816974147.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56299#issuecomment-2450999471) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3e0825e_vs_7715cf2/report.md).

Keno committed:
Move jl_get_globalref_partition to Julia for performance

Keno commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56299#issuecomment-2451086100) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/8ac81f7_vs_7715cf2/report.md).

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -2316,6 +2316,191 @@ function abstract_throw_methoderror(interp::AbstractInterpreter, argtypes::Vecto
+            # rt = Const(nothing)
+            rt = Type
+        elseif is_some_const_binding(binding_kind(partition))
+            rt = Const(Any)
```
Why not returning the type of the constant?

aviatesk replied:
Ah, maybe https://github.com/JuliaLang/julia/blob/9d1cedb2379111a01eab53b811fbf76a1af53bd2/src/module.c#L610-L615

Keno replied:
yes

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -3063,45 +3285,116 @@ function override_effects(effects::Effects, override::EffectsOverride)
+            force_binding_resolution!(g)
+            return abstract_eval_globalref_type(g, src, false)
+        end
+        # return Union{}
```
What's the TODO remaining to return `Union{}` here?

aviatesk replied:
Ah, I guess `# We do not currently assume an invalidation for guard -> defined transitions`?

Keno replied:
yes

aviatesk commented on base/compiler/tfuncs.jl:
```diff
@@ -2525,11 +2495,13 @@ function builtin_effects(𝕃::AbstractLattice, @nospecialize(f::Builtin), argty
-        effect_free = get_binding_type_effect_free(argtypes[1], argtypes[2]) ? ALWAYS_TRUE : ALWAYS_FALSE
-        return Effects(EFFECTS_TOTAL; effect_free)
+        # Modeled more precisely in abstract_eval_get_binding_type
+        return Effects(EFFECTS_TOTAL; effect_free=ALWAYS_FALSE)
```
`nothrow=false` instead of `effect_free=ALWAYS_FALSE`? This code path might not be too important though.

Keno commented on base/runtime_internals.jl:
```diff
@@ -230,13 +230,20 @@ const BINDING_KIND_DECLARED     = 0x7
+    if isdefined(gr, :binding)
+        b = gr.binding
+    else
+        b = ccall(:jl_get_module_binding, Ref{Core.Binding}, (Any, Any, Cint), gr.mod, gr.name, true)
```
That definition was probably wrong, but regardless, this code can't handle the binding not existing.
